### DISCUSSION TITLE
Live Emote Updates

### DISCUSSION THREAD

**Discussion by Nerixyz** [Chatterino/chatterino2#4038] [Ideas]
7TV and BTTV both have live emote updates in their web extensions and 7TV has it in their fork of Chatterino as well.

I mentioned in #4002 a reason to add 7TV's updates. Since all platforms should be treated equally either both or none should be added.

### References

* 7TV: [SevenTV/EventApi](https://github.com/SevenTV/EventAPI/), SevenTV/Chatterino7#138 (initial implementation)
* BTTV: night/betterttv#5630 (no real documentation)

**Comment by leon-richardt** [comment]
For the sake of completeness, here is the reason outlined in #4002:
> I noticed, in contrast to BTTV and FFZ, users tend to add/remove/update emotes more frequently and having used it I must say it's very nice not having to press F5 every minute in some channels. In addition, the [user-by-connection](https://github.com/SevenTV/API/blob/a84e884b5590dbb5d91a5c6b3548afabb228f385/internal/rest/v3/routes/users/users.by-connection.go#L28) endpoint caches the response for 60s so F5 might not work as expected immediately.

**Comment by leon-richardt** [comment]
It's worth noting that we had some discussion on this feature on stream recently.
@pajlada suggested that we should consider proxying live emote updates through the Chatterino API (or potentially a new service, hosted by us). The idea is to provide a uniform interface for the Chatterino client so that implementations for different emote providers can be handled through a common interface. Defining our own interface would also allow us to handle changes in the upstream APIs without having to release a new client version. Furthermore, this would move load away from the upstream servers and onto our own. (Although traffic from Chatterino is likely negligible compared to the traffic generated by the respective web extensions.)

I'm sure there are more good reasons I can't think of right now; feel free to mention them as well.

**Comment by Nerixyz** [comment]
Since there's an existing, functional implementation for 7TV's EventApi in Chatterino7, I went ahead and added a BTTV implementation as well. Now they both share a common superclass `BasicPubSubClient`, since they are very similar (except that BTTV doesn't have/require a heartbeat).

I'm not sure if I support the idea of creating a separate service for emote-updates. This would essentially require a similar effort on the Chatterino side (creating a new WebSocket(-pool), updating emote-maps...) and a lot of additional effort to create and run the external service. The main advantage of the additional service would be that Chatterino could be partially (or with a greater effort _entirely(?)_) decoupled from emote-providers. But the API of these emote-providers seems pretty stable, thus the liveupdates should be stable [enough] as well. 

I'm fine with opening a (draft) PR for code review and splitting the implementation up.

* [Branch `experiment/live-updates`](https://github.com/Nerixyz/chatterino2/tree/experiment/live-updates)
* [Diff to master](https://github.com/Chatterino/chatterino2/compare/master...Nerixyz:chatterino2:experiment/live-updates)

**Reply by pajlada** [reply]
Repeating what I said in chat; with 2 of 4 emote providers providing their own API
 - [X] 7TV
 - [X] BetterTTV
 - [ ] FrankerFaceZ
 - [ ] Twitch

I'm fine with just connecting to each of their Event APIs and getting live updates from there.
What's important to me is ensuring we have permission to implement & use their APIs, whether it's from their developer agreement or asking in a GitHub issue/twitter before we merge any such feature into the main branch.

Another thing that's important is to understand that the APIs will probably not be stable, so let's plan for resilience and ensure we test for scenarios like failed connections, disconnects, repeated disconnects, changed schemas, receiving random messages we never asked for etc.

**Comment by Felanbird** [comment]
Both 7TV and BTTV live-emote-updates were added in some PRs I'm not going to look for, if FFZ adds live updating in the future we can open a new discussion.
